### PR TITLE
parallel dereference.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @ansyral @hellosnow @vicancy @qinezh @superyyrrzz

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 1]
```diff
@@ -193,29 +194,27 @@ public static void Dereference(this Manifest manifest, string manifestFolder)
...
-            }
+                    try
+                    {
+                        fal.Copy(ofi.RelativePath, ofi.RelativePath);
+                        ofi.LinkToPath = null;
+                    }
+                    catch (PathTooLongException)
+                    {
+                        Logger.LogError($"Unable to dereference file: {ofi.RelativePath}.", file: ofi.RelativePath);
+                    }
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: throw?

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 2]
```diff
@@ -193,29 +194,27 @@ public static void Dereference(this Manifest manifest, string manifestFolder)
...
-                foreach (var rp in ofiList)
-                {
-                    fal.Copy(rp.RelativePath, rp.RelativePath);
-                }
-                foreach (var ofi in ofiList)
+            var fal = FileAbstractLayerBuilder.Default
+                .ReadFromManifest(manifest, manifestFolder)
+                .WriteToRealFileSystem(manifestFolder)
+                .Create();
+            Parallel.ForEach(
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: reuse MaxParallelism?

**Comment by hellosnow** [comment]
>             foreach (var g in from m in items

need we parallel shrink?

---
Refers to: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs:230 in d20674d. [](commit_id = d20674d4f0c8e819397552089ab83bba4e7152f8, deletion_comment = False)

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 3]
```diff
@@ -193,29 +194,27 @@ public static void Dereference(this Manifest manifest, string manifestFolder)
...
-                }
-            }
+                    try
+                    {
+                        fal.Copy(ofi.RelativePath, ofi.RelativePath);
+                        ofi.LinkToPath = null;
+                    }
+                    catch (PathTooLongException)
+                    {
+                        Logger.LogError($"Unable to dereference file: {ofi.RelativePath}.", file: ofi.RelativePath);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: >Unable [](start = 42, length = 6)

minor: add some exception info in the message?

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 4]
```diff
@@ -9,6 +9,7 @@ namespace Microsoft.DocAsCode.Common
     using System.Linq;
 
     using Microsoft.DocAsCode.Plugins;
+    using System.Threading.Tasks;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: order

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 5]
```diff
@@ -182,7 +183,7 @@ public static T Modify<T>(this Manifest manifest, Func<Manifest, T> func)
             }
         }
 
-        public static void Dereference(this Manifest manifest, string manifestFolder)
+        public static void Dereference(this Manifest manifest, string manifestFolder, int parallism)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: >parallism [](start = 90, length = 9)

typo: parallelism

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs [hunk 6]
```diff
@@ -193,57 +194,59 @@ public static void Dereference(this Manifest manifest, string manifestFolder)
...
+                new ParallelOptions { MaxDegreeOfParallelism = parallism > 0 ? parallism : Environment.ProcessorCount },
+                list =>
                 {
-                    File.Delete(Environment.ExpandEnvironmentVariables(item.LinkToPath));
-                    item.LinkToPath = first.LinkToPath;
-                }
-            }
+                    foreach (var item in list.Skip(1))
+                    {
+                        File.Delete(Environment.ExpandEnvironmentVariables(item.LinkToPath));
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: >item.LinkToPath [](start = 75, length = 15)

distinct item.LinkToPath before deleting?

**Comment by ansyral** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs (hunk 4 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/FileAbstractLayer/ManifestFileHelper.cs]: >using System.Threading.Tasks; [](start = 4, length = 29)

order

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**